src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -1369,7 +1374,7 @@ static const uint32_t *dds_stream_read_seq (dds_istream_t * __restrict is, char
   dds_sequence_t * const seq = (dds_sequence_t *) addr;
   const enum dds_stream_typecode subtype = DDS_OP_SUBTYPE (insn);
   uint32_t bound_op = seq_is_bounded (DDS_OP_TYPE (insn)) ? 1 : 0;
-  if (subtype > DDS_OP_VAL_8BY && is->m_xcdr_version == CDR_ENC_VERSION_2)
+  if (!is_primitive_type (subtype) && is->m_xcdr_version == CDR_ENC_VERSION_2)
```

Comment:
Like the primitive-or-enum check, this one also occurs fairly often, and it seems like it could be replaced by a `is_dheader_needed` function. If it works for at least the vast majority of the cases, it think it's worth it.

Reply:
Good suggestion, I've added both `is_primitive_or_enum_type` and `is_dheader_needed`